### PR TITLE
improve folder/file terminal completions

### LINKED ISSUE #234351: Have `.` and `..` as completions when completing directories
`.` (and `..`) are valid folders that we should provide as a completion for directories.

See pwsh support:

Image

Image

Image

When there's a directory before it, `..` is there with a trialing `/` or `\`:

Image

### PR DESCRIPTION
**Pull request by meganrogge**
fixes #234351 and some other bugs (description not showing up for suggestions etc)

Also enables this for pwsh 

Adds `cd` completion spec and refactors to pull out `arg` handling

### CODE REVIEW

File: extensions/terminal-suggest/src/completions/cd.ts [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+const cdSpec: Fig.Spec = {
+	name: 'cd',
+	description: 'Change the shell working directory',
+	args: {
+		name: 'folder',
+		template: 'folders',
+		isVariadic: true,
+
+		// Add an additional hidden suggestion so users can execute on it if they want to
```
Comment (meganrogge) [inline extensions/terminal-suggest/src/completions/cd.ts:14]:
```suggestion
```

File: extensions/terminal-suggest/src/completions/cd.ts [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+		suggestions: [
+			{
+				name: '-',
+				description: 'Switch to the last used folder',
+				hidden: true,
+			},
```
Comment (Tyriar) [inline extensions/terminal-suggest/src/completions/cd.ts:19]: Created https://github.com/microsoft/vscode/issues/234382 to tweak this for pwsh

File: src/vs/workbench/contrib/terminalContrib/suggest/browser/terminalCompletionService.ts [hunk 3]
```diff
@@ -197,44 +197,47 @@ export class TerminalCompletionService extends Disposable implements ITerminalCo
...
-				isFile: kind === TerminalCompletionItemKind.File,
-				replacementIndex: cursorPosition,
-				replacementLength: label.length
-			});
+			for (const stat of fileStat.children) {
+				let kind: TerminalCompletionItemKind | undefined;
+				if (foldersRequested && stat.isDirectory) {
+					kind = TerminalCompletionItemKind.Folder;
+				}
+				if (filesRequested && !stat.isDirectory && (stat.isFile || stat.resource.scheme === 'file')) {
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminalContrib/suggest/browser/terminalCompletionService.ts:222]: `Schemas.file` from network.ts is preferable.

File: extensions/terminal-suggest/src/terminalSuggestMain.ts [hunk 4]
```diff
@@ -106,26 +107,17 @@ export async function activate(context: vscode.ExtensionContext) {
...
-					uniqueResults.set(item.label, item);
-				}
-			}
-			const resultItems = uniqueResults.size ? Array.from(uniqueResults.values()) : undefined;
-
-			// If no completions are found, the prefix is a path, and neither files nor folders
+			// If the command line is empty, no completions are found, or the completion found is '.', the prefix is a path, and neither files nor folders
 			// are going to be requested (for a specific spec's argument), show file/folder completions
-			const shouldShowResourceCompletions = !resultItems?.length && prefix.match(/^[./\\ ]/) && !filesRequested && !foldersRequested;
+			const shouldShowResourceCompletions = (terminalContext.commandLine.trim().length === 0 || !items?.length || items.length === 1 && items[0].label === '.') && (!filesRequested && !foldersRequested);
```
Comment (Tyriar) [inline extensions/terminal-suggest/src/terminalSuggestMain.ts:112]: nit: you could move the comments closer to the conditions for harder to parse code like this:

```
const shouldShowResourceCompletions = (
  // Command line is empty
  terminalContext.commandLine.trim().length === 0 ||
  // ...
  !items?.length ||
  ...
);
```

**Review by Tyriar — DISMISSED** [review]

File: src/vs/workbench/contrib/terminalContrib/suggest/browser/terminalCompletionService.ts (hunk 3 above)
Comment (meganrogge) [reply src/vs/workbench/contrib/terminalContrib/suggest/browser/terminalCompletionService.ts:222]: yeah, tried that yesterday for awhile and auto import was totally broken. works today

Comment (meganrogge) [reply src/vs/workbench/contrib/terminalContrib/suggest/browser/terminalCompletionService.ts]: yeah, tried that yesterday for awhile and auto import was totally broken. works today

**Review by Tyriar — APPROVED** [review]

**Pull request merged**